src/base/Identifier.ml

```diff
@@ -19,21 +19,89 @@
 open Core_kernel
 open ErrorUtils
 
-type 'rep t = Ident of string * 'rep [@@deriving sexp]
+module type QualifiedName = sig
 
-let asId i = Ident (i, dummy_loc)
+  type t [@@deriving sexp]
 
-let asIdL i loc = Ident (i, loc)
+  val as_string : t -> string
 
-let get_id i = match i with Ident (x, _) -> x
+  val as_error_string : t -> string
 
-let get_rep i = match i with Ident (_, l) -> l
+  val equal_id : t -> t -> bool
+  
+  val compare_id : t -> t -> int
+    
+end
+
+module FlattenedName = struct
```

Comment:
I'm wondering if it makes sense here to restrict the type of FlattenedName like so
```
module FlattenedName : QualifiedName = struct
```
It seems that it should provide some more type-safety: it would be impossible to mix strings and identifiers in some cases unless I'm missing something.

Reply:
This only provides more type safety when `Name.t` is `string`. Once `Name.t` becomes an ADT, then there is no risk of confusing identifiers and strings, and by then the priority is for the constructor names to be visible, which the above suggestion will prevent.